yrs/src/id_set.rs

```diff
@@ -1,23 +1,200 @@
 
 use crate::*;
-use crate::block::Block;
+use crate::block::{Block, ID};
+use crate::updates::encoder::{EncoderV1, UpdateEncoder, DSEncoder};
+use crate::updates::decoder::{DecoderV1, UpdateDecoder, DSDecoder};
 
-#[derive(Default)]
-struct IdRange {
-    clock: u32,
-    len: u32
+#[derive(Default, Copy, Clone)]
+pub(crate) struct IdRange {
+    pub clock: u32,
+    pub len: u32
+}
+
+impl IdRange {
+    pub fn new(clock: u32, len: u32) -> Self {
+        IdRange {
+            clock,
+            len,
+        }
+    }
+
+    pub fn encode(&self, encoder: &mut EncoderV1) {
+        encoder.write_ds_clock(self.clock);
+        encoder.write_ds_len(self.len);
+    }
+
+    pub fn decode(decoder: &mut DecoderV1) -> Self {
+        let clock = decoder.read_ds_clock();
+        let len = decoder.read_ds_len();
+        IdRange {
+            clock,
+            len,
+        }
+    }
 }
 
+/// DeleteSet is a temporary object that is created when needed.
+/// - When created in a transaction, it must only be accessed after sorting and merging.
+///   - This DeleteSet is sent to other clients.
+/// - We do not create a DeleteSet when we send a sync message. The DeleteSet message is created
+///   directly from StructStore.
+/// - We read a DeleteSet as a apart of sync/update message. In this case the DeleteSet is already
+///   sorted and merged.
 #[derive(Default)]
 pub struct IdSet {
     clients: HashMap::<u64, Vec<IdRange>, BuildHasherDefault<ClientHasher>>,
 }
 
+pub(crate) type Iter<'a> = std::collections::hash_map::Iter<'a, u64, Vec<IdRange>>;
+
 impl IdSet {
     pub fn new () -> Self {
         Self::default()
     }
-    pub fn iterate_blocks (&self, tr: &mut Transaction, f: fn(Block)) {
 
+    pub fn from(store: &BlockStore) -> Self {
+        let mut set = Self::new();
+        for (&client, blocks) in store.clients.iter() {
+            let mut ranges = Vec::with_capacity(blocks.list.len());
+            let mut i = 0;
+            while i < blocks.list.len() {
+                let block = &blocks.list[i];
+                if block.is_deleted() {
+                    let clock = block.id().clock;
+                    let mut len = block.len();
+                    if i + 1 < blocks.list.len() {
+                        let mut next = &blocks.list[i+1];
+                        while i + 1 < blocks.list.len() && next.id().clock == clock + len && next.is_deleted() {
```

Comment:
I think we can remove this condition `next.id().clock == clock + len`. This should always hold true.

Reply:
Had it ever failed in y.js? I've ported it from there. Usually there's a reason for such condition to happen in first place.

Reply:
Previous versions of Yjs didn't always require that all operations are present. So you could have `id={clock: 1}` but `id={clock: 0}` could still be missing. I think this is reminiscent. 

I just removed it from Yjs and the test suite is still running fine. 
https://github.com/yjs/yjs/commit/4e343ccace9dca0b594ab42d5934b43e95d245c6

The CI that runs the complete test suite is still running. I'll ping here if it breaks.